core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule.fuse;
+
+import com.google.common.collect.ImmutableSet;
+import io.trino.Session;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.AggregationNode;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.tree.SymbolReference;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.trino.SystemSessionProperties.isFuseSubPlanEnabled;
+import static io.trino.sql.planner.iterative.rule.fuse.PlanNodeFuser.outputsSameAsSource;
+import static io.trino.sql.planner.plan.Patterns.join;
+import static io.trino.sql.tree.BooleanLiteral.TRUE_LITERAL;
+
+/**
+ * Simplified "Computation Reuse via Fusion in Amazon Athena" paper JoinOnKeys rule
+ * that works only on a single cross join over global aggregations over the same table.
+ * It transforms
+ * <pre>
+ *   cross join
+ *     aggregation global
+ *       agg_left <- aggLeft()
+ *         filter(filter_left)
+ *           table scan(table)
+ *     aggregation global
+ *       agg_right <- aggRight()
+ *         filter(filter_right)
+ *           table scan(table)
+ * </pre>
+ * into:
+ * <pre>
+ *   aggregation global
+ *       agg_left <- aggLeft(mask = agg_mask_left)
+ *       agg_right <- aggRight(mask = agg_mask_right)
+ *     project(agg_mask_left = filter_left, agg_mask_left = filter_right)
+ *       filter(filter_left or filter_right)
+ *         table scan
+ * </pre>
+ * <p>
+ * This makes the plan to read the source table only once,
+ * resulting in the potentially significant performance improvement.
+ */
+public class FuseCrossJoinedGlobalAggregations
+        implements Rule<JoinNode>
+{
+    private static final Pattern<JoinNode> PATTERN = join()
+            .matching(JoinNode::isCrossJoin);
```

Comment:
Do you mean instead of 
```
PlanNode left = context.getLookup().resolve(node.getLeft());
PlanNode right = context.getLookup().resolve(node.getRight());
``` 
?

Reply:
done